execute/table.go

```diff
@@ -36,43 +36,80 @@ func GroupKeyForRowOn(i int, cr flux.ColReader, on map[string]bool) flux.GroupKe
 	return NewGroupKey(cols, vs)
 }
 
-// OneTimeTable is a Table that permits reading data only once.
-// Specifically the ValueIterator may only be consumed once from any of the columns.
-type OneTimeTable interface {
-	flux.Table
-	onetime()
-}
-
-// CacheOneTimeTable returns a table that can be read multiple times.
-// If the table is not a OneTimeTable it is returned directly.
-// Otherwise its contents are read into a new table.
-func CacheOneTimeTable(t flux.Table, a *memory.Allocator) (flux.Table, error) {
-	_, ok := t.(OneTimeTable)
-	if !ok {
-		return t, nil
-	}
-	return CopyTable(t, a)
+type tableBuffer struct {
+	key     flux.GroupKey
+	colMeta []flux.ColMeta
+	i       int
+	buffers []flux.ColReader
 }
 
-// CopyTable returns a copy of the table and is OneTimeTable safe.
-func CopyTable(t flux.Table, a *memory.Allocator) (flux.Table, error) {
-	builder := NewColListTableBuilder(t.Key(), a)
+func (tb *tableBuffer) Key() flux.GroupKey {
+	return tb.key
+}
 
-	cols := t.Cols()
-	colMap := make([]int, len(cols))
-	for j, c := range cols {
-		colMap[j] = j
-		if _, err := builder.AddCol(c); err != nil {
-			return nil, err
+func (tb *tableBuffer) Cols() []flux.ColMeta {
+	return tb.colMeta
+}
+
+func (tb *tableBuffer) Do(f func(flux.ColReader) error) error {
+	defer tb.Done()
```

Comment:
By deferring Done, don't you Release twice?

Reply:
`Done` is idempotent. The iteration calls `Release` as it processes tables, but it keeps the current index so it won't iterate over that section again.